File: LinkedList/linked_list.py
class Node:
    def __init__(self, data, pnext: 'Node'=None):
        self.data  = data
        self.pnext = pnext

    def display_node(self):
        print(f'{self.data} ->', end=' ')


class SimpleLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def is_empty(self) -> bool:
        """Return FALSE if linked list is EMPTY"""
        return self.head is None
    
    def add_head(self, x):
        """Add x to head of linked list"""
        node = Node(x)
        if self.is_empty():
            self.head  = node
            self.tail  = node
        else:
            node.pnext = self.head
            self.head  = node

    def add_tail(self, x):
        """Add x to tail of linked list"""
        node = Node(x)
        if self.is_empty():
            self.head       = node
            self.tail       = node
        else:
            self.tail.pnext = node
            self.tail       = node

    def insert(self, x, index: int):
        """Insert x to index position"""

        if self.is_empty() or index == 0:
            self.add_head(x)
            return
        elif index > self.count_node():
            print(f'Error: index is out of range: {index} > {self.count_node()}')
            return
        
        pindex   = 0        
        pcurrent = self.head       
        
        while pindex != index - 1:
            pcurrent  = pcurrent.pnext
            pindex   += 1       
        
        node           = Node(x, pcurrent.pnext)
        pcurrent.pnext = node

        if pcurrent == self.tail:
            self.tail = node        

    def remove_index(self, index: int):
        """Remove value at index"""
        if self.is_empty() or index >= self.count_node():
            return
        pindex   = 0
        pcurrent = self.head

        while pindex  != index - 1:
            pcurrent   = pcurrent.pnext
            pindex    += 1
        pcurrent.pnext = pcurrent.pnext.pnext
    
    def remove_head(self):
        """Remove head"""
        if self.is_empty():
            return
        elif self.head == self.tail:
            del self.head
            del self.tail
            self.head = None
            self.tail = None
        else:
            self.head = self.head.pnext

    def remove_tail(self):
        """Remove tail"""
        if self.is_empty():
            return
        elif self.head == self.tail:
            self.head = None
            self.tail = None
        else: 
            pcurrent = self.head
            while pcurrent.pnext != self.tail:
                pcurrent = pcurrent.pnext
                
            del self.tail
            pcurrent.pnext = None

    def count_node(self) -> int:
        """Return number of nodes"""
        count = 0
        if self.is_empty():
            return 0
        pcurrent = self.head
        while pcurrent:
            count += 1
            pcurrent = pcurrent.pnext
        return count

    def sum_node(self) -> int:
        """Return Sum nodes"""
        nsum = 0
        if self.is_empty():
            return 0
        pcurrent = self.head
        while pcurrent:
            nsum    += pcurrent.data
            pcurrent = pcurrent.pnext
        return nsum

    def max_node(self) -> int:
        """Return max node in linked list"""
        nmax = self.head.data
        if self.is_empty():
            return 0
        pcurrent = self.head.pnext
        while pcurrent:
            if pcurrent.data > nmax:
                nmax = pcurrent.data
            pcurrent = pcurrent.pnext
        return nmax

    def update_node(self, x, index: int=0):
        """Update node value"""
        pindex = 0
        if self.is_empty():
            return
        pcurrent = self.head
        while pcurrent and pindex != index:
            pcurrent  = pcurrent.pnext
            pindex   += 1

        if pcurrent:
            pcurrent.data = x

    def travesal(self):
        if self.is_empty():
            print('LinkedList is Empty')
        pcurrent = self.head
        while pcurrent:
            pcurrent.display_node()
            pcurrent = pcurrent.pnext

    def __iter__(self):
        pcurrent = self.head
        while pcurrent:
            yield pcurrent.data
            pcurrent = pcurrent.pnext

#  DEMO
link_list = SimpleLinkedList()

link_list.add_head(1)
link_list.add_tail(2)
link_list.add_tail(3)
link_list.travesal()
print()
link_list.insert(56, 3)
link_list.travesal()
print()
print(link_list.tail.data)

File: LinkedList/linked_list.cpp
#include <iostream>
#include <list>
#include <utility>

struct Node {
    int          data;
    struct Node* next;
    Node(int d, Node* n=nullptr) : data(d), next(n) {}
};

class LinkedList {
    public:
        Node* head = nullptr;
        Node* tail = nullptr;
    
    bool is_empty() {
        return this->head == nullptr;
    }

    void push_front(int x) {
        Node* new_node = new Node(x, nullptr);
        if (is_empty()) {
            this->head = new_node;
            this->tail = new_node;
        } else {
            new_node->next = this->head;
            this->head = new_node;
        }
    }
    void push_back(int x) {
        Node* new_node = new Node(x, nullptr);
        if (is_empty()) {
            this->head = new_node;
            this->tail = new_node;
        } else {
            this->tail->next = new_node;
            this->tail       = new_node;
        }
    }
    void remove_head() {
        if (is_empty()) {
            return;
        } else if (this->head == this->tail) {
            delete this->head;
            this->head = this->tail = nullptr;
        } else {
            Node* temp = this->head;
            this->head = this->head->next;
            delete temp;
        }
    }
    void remove_tail() {
        if (is_empty()) {
            return;
        } else if (this->head == this->tail) {
            delete this->head;
            this->head = this->tail = nullptr;
        } else {
            Node* pcurrent = this->head;
            while (pcurrent->next != this->tail) {
                pcurrent = pcurrent->next;
            }
            delete this->tail;
            pcurrent->next = nullptr;
        }
    }
    void remove_index(int index=0) {
        if (is_empty()) {
            return;
        }
        if (index == 0) {
            remove_head();
        }
        Node* pcurrent = this->head;
        int   pindex   = 0;
        while (pcurrent->next != nullptr && pindex + 1 < index) {
            pcurrent = pcurrent->next;
            pindex++;
        }
        Node* target = pcurrent->next;
        if (target == nullptr) return;
        
        if (pcurrent->next == this->tail) {
            this->tail = pcurrent;
        }
        pcurrent->next = target->next;
        delete target;
    }
    void update_node(int data, int index) {
        if (is_empty()) {
            return;
        }
        int   pindex   = 0;
        Node* pcurrent = this->head;
        while (pcurrent && pindex < index) {
            pcurrent = pcurrent->next;
            pindex ++;
        }
        if (pcurrent) {
            pcurrent->data = data;
        }
    }
    int length() {
        int c = 0;
        Node *pcurrent = this->head;
        while (pcurrent != nullptr) {
            pcurrent = pcurrent->next;
            c++;
        }
        return c;
    }
    void travesal() {
        Node* pcurrent = this->head;
        std::cout << "--Travesal--" << std::endl;
        while (pcurrent != nullptr) {
            Node* next = pcurrent->next;
            std::cout << pcurrent->data << " -> ";
            pcurrent = next;
        }
        std::cout << std::endl;
    };    
    ~LinkedList() {
        Node* pcurrent = this->head;
        while (pcurrent != nullptr) {
            Node* next = pcurrent->next;
            delete pcurrent;
            pcurrent   = next;
        }
    }
};

int main() {
    // std::list<int> linked_list;
    // linked_list.push_back(0);
    // linked_list.push_front(5);
    // linked_list.push_back(10);
    // linked_list.push_front(20);
    // linked_list.push_front(25);
    // for (const int& element : linked_list) {
    //     std::cout  << element << " -> ";
    // 

    LinkedList my_linked_list;
    my_linked_list.push_back(0);
    my_linked_list.push_front(5);
    my_linked_list.push_back(10);
    my_linked_list.push_front(20);
    my_linked_list.push_front(25);
    my_linked_list.remove_head();
    my_linked_list.remove_tail();
    my_linked_list.remove_index(2);
    my_linked_list.update_node(3, 1);
    my_linked_list.travesal();
    std::cout << "Length: " << my_linked_list.length() << std::endl;
}